lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp

```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+    // expected: The default lldb_expr wrapper produces some warnings.
+    if (!m_num_clang_errors || m_num_swift_errors)
+      for (const RawDiagnostic &diagnostic : m_raw_swift_diagnostics)
+        diagnostic_manager.AddDiagnostic(diagnostic.description.c_str(),
+                                         SeverityForKind(diagnostic.kind),
+                                         eDiagnosticOriginSwift);
+  }
+
+  bool GetColorize() const { return m_colorize; }
+
+  bool SetColorize(bool b) {
+    const bool old = m_colorize;
+    m_colorize = b;
+    return old;
+  }
+
+  void AddDiagnostic(std::unique_ptr<Diagnostic> diagnostic) {
+    if (diagnostic)
+      m_diagnostics.push_back(std::move(diagnostic));
+  }
+
+private:
+  // We don't currently use lldb_private::Diagostic or any of the lldb
+  // DiagnosticManager machinery to store diagnostics as they
+  // occur. Instead, we store them in raw form using this struct, then
+  // transcode them to SwiftDiagnostics in PrintDiagnostic.
+  struct RawDiagnostic {
+    RawDiagnostic() = default;
+    RawDiagnostic(std::string in_desc, swift::DiagnosticKind in_kind,
+                  llvm::StringRef in_bufferName, unsigned in_bufferID,
+                  uint32_t in_line, uint32_t in_column,
+                  llvm::ArrayRef<swift::Diagnostic::FixIt> in_fixits)
+        : description(in_desc), kind(in_kind), bufferName(in_bufferName),
+          bufferID(in_bufferID), line(in_line), column(in_column),
+          fixits(in_fixits) {}
+    RawDiagnostic(const RawDiagnostic &other)
+        : description(other.description), kind(other.kind),
+          bufferName(other.bufferName), bufferID(other.bufferID),
+          line(other.line), column(other.column), fixits(other.fixits) {}
+    bool operator==(const RawDiagnostic& other) {
+      return kind == other.kind && bufferID == other.bufferID &&
+             line == other.line && column == other.column &&
+             description == other.description;
+    }
+    bool operator!=(const RawDiagnostic &other) { return !(*this == other); }
+
+    std::string description;
+    swift::DiagnosticKind kind = swift::DiagnosticKind::Error;
+    llvm::StringRef bufferName;
+    unsigned bufferID;
+    uint32_t line;
+    uint32_t column;
+    std::vector<swift::DiagnosticInfo::FixIt> fixits;
+  };
+
+  SwiftASTContext &m_ast_context;
+  /// Stores all diagnostics coming from the Swift compiler.
+  std::vector<RawDiagnostic> m_raw_swift_diagnostics;
+  std::vector<RawDiagnostic> m_raw_clang_diagnostics;
+  /// Stores diagnostics coming from LLDB.
+  std::vector<std::unique_ptr<Diagnostic>> m_diagnostics;
+
+  unsigned m_num_swift_errors = 0;
+  unsigned m_num_clang_errors = 0;
+  bool m_colorize = false;
+
+  friend class SwiftASTContext::TransactionDiagnostics;
+};
+
+} // namespace lldb_private
+
+SwiftASTContext::TransactionDiagnostics::TransactionDiagnostics(
+    swift::DiagnosticConsumer &consumer)
+    : m_consumer(consumer),
```

Comment:
You could make `m_consumer` a `StoringDiagnosticConsumer` and then you'd only have to do the cast oncee.

Reply:
I'm trying to not introduce any Swift compiler dependencies into SwiftASTContext.h